lambda_src/lambda_function.py

```diff
@@ -97,15 +102,115 @@ def sync_flow(event: Any, context: Any = None) -> Dict[Text, Any]:
     LOG.debug('Destination header not found in a POST and hence using sync_flow().')
     headers = event['headers']
     req_body = loads(event['body'])
+    start_time = timer()
 
+    destination_driver = None
     batch_id = headers[BATCH_ID_HEADER]
     write_uri = headers.get('write-uri')
+
     LOG.debug(f'sync_flow() received destination: {write_uri}.')
 
     if write_uri:
         destination_driver = import_module(
             f'geff.drivers.destination_{urlparse(write_uri).scheme}'
         )
+
+    while True:
+        data = get_data_from_lock(batch_id)
+        if data is None:  # request hasn't been initialized
+            break
+        elif data != LOCKED:  # if false, the response is yet to be written
+            return data
+
+    open_lock(batch_id)  # initilaze the request
+
+    res_data = process_request(
+        req_body, headers, event, batch_id, write_uri, destination_driver
+    )
+
+    # Write data to s3 or return data synchronously
+    if write_uri:
+        response = destination_driver.finalize(  # type: ignore
+            write_uri, batch_id, res_data
+        )
+    else:
+        data_dumps = dumps({'data': res_data}, default=str)
+        response = {
+            'statusCode': 200,
+            'body': b64encode(compress(data_dumps.encode())).decode(),
+            'isBase64Encoded': True,
+            'headers': {'Content-Encoding': 'gzip'},
+        }
+        end_time = timer()
+        if response and (end_time - start_time) > 20:
+            LOG.debug('Storing the response in DynamoDB.')
+            try:
+                close_lock(batch_id, response)  # write the response
+            except ClientError as ce:
+                if ce.response['Error']['Code'] == 'ValidationException':
+                    LOG.error(ce)
+                    pass
+
+    if len(response) > 6_000_000:
+        response = response_size_error(response, req_body)
+    return response
+
+
+def response_size_error(
+    response_input: Dict[Text, Any], req_body: Dict[Text, Any]
+) -> str:
+    """
+    Creates a new response object with an error message,
+    for when the response size is likely to exceed the allowed payload size.
+
+    Args:
+        response (Dict[Text, Any]): Response object to calculate the size.
```

Comment:
`response_input` seems like an undescriptive name here -- it's input because it's the arg to a function?

it's a string, not a dict.

Reply:
> response_input seems like an undescriptive name here -- it's input because it's the arg to a function?

Was thinking something along the lines of that, will think of a more descriptive name.

> it's a string, not a dict.

Sorry, but it seems like it's a dict? ([1](https://github.com/Snowflake-Labs/geff/blob/e601e97a1aa90649ebc18f15a66cae1a1b8e8a6b/lambda_src/lambda_function.py#L154), [2](https://github.com/Snowflake-Labs/geff/blob/2aa38df9a4345e40e53a6ee4dc261dae66be3398/lambda_src/drivers/destination_s3.py#L133))

Reply:
The one that returns `statusCode` is the correct one to return to the lambda runtime. this function returns a string, and `finalize` returns a dictionary that is not properly formatted and so should be erroring when the async call completes. the best way to fix this is probably by using [TypedDict](https://docs.python.org/3.8/library/typing.html#typing.TypedDict). up to you if you want to do that refactoring here, tho -- keeping things inline as they are seems fine, as well.